docs/proposals/llm-hyperparameter-optimization-api.md

```diff
@@ -0,0 +1,191 @@
+  * [Links](#links)
+  * [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+  * [Design for API](#design-for-api)
+    + [Example](#example)
+  * [Implementation](#implementation)
+
+## Links
+
+- [katib/issues#2291 (Tuning API in Katib for LLMs)](https://github.com/kubeflow/katib/issues/2291)
+
+## Motivation
+
+The rapid advancements and growing popularity of Large Language Models (LLMs) have driven an increased need for effective LLMOps in Kubernetes environments. To address this, we developed a [train API](https://www.kubeflow.org/docs/components/training/user-guides/fine-tuning/) within the Training Python SDK, simplifying the process of fine-tuning LLMs using distributed PyTorchJob workers. However, hyperparameter optimization remains a crucial yet labor-intensive task for enhancing model performance. Automating this tuning process through a dedicated API will facilitate efficient and scalable exploration of hyperparameters, ultimately improving model performance and reducing manual effort.
+
+## Goals
+
+Our goal is to develop a high-level API for tuning hyperparameters of LLMs that simplifies the process of hyperparameter optimization in Kubernetes. This API will seamlessly integrate with external platforms like HuggingFace and S3 for importing pretrained models and datasets. By specifying parameters for the training objective, trial configurations, and PyTorch worker configurations, the API will automate the creation of experiments and execution of trials. This abstraction of Kubernetes infrastructure complexities will enable data scientists to optimize hyperparameters efficiently and effectively.
+
+## Non-Goals
+
+1. Incorporate early stopping strategy into the API to optimize training efficiency.
+2. Expand support for distributed training in frameworks beyond PyTorch by leveraging their distributed training capabilities.
+3. Support adding custom providers through configmap or CRD approach to enhance flexibility.
+4. Enable users to deploy tuned models for inference within their applications or seamlessly integrate them into existing NLP pipelines for specialized tasks.
+
+## Design for API
+
+![Design for API](images/design_api.jpg)
+
+```python
+import kubeflow.katib as katib
+from kubeflow.katib import KatibClient
+
+class KatibClient(object):
+	
+	def tune(
+		name: str, 
+		namespace: Optional[str] = None,
+		model_provider_parameters = None,
+		dataset_provider_parameters = None,
+		trainer_parameters = None,
+		algorithm_name: str = "random",
+		algorithm_settings: Union[dict, List[models.V1beta1AlgorithmSetting], None] = None,
+		objective_metric_name: str = "eval_accuracy",
+		additional_metric_names: List[str] = [],
+		objective_type: str = "maximize",
+		objective_goal: float = None,
+		max_trial_count: int = None,
+		parallel_trial_count: int = None,
+		max_failed_trial_count: int = None,
+		resources_per_trial = {
```

Comment:
I changed the name from "pytorch_config" to "resources_per_trial" and used a dictionary to set the parameters for PyTorchJob. Please take a look. @andreyvelich @johnugeorge

Reply:
To be compatible with [the existing API](https://github.com/kubeflow/katib/blob/master/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L185C9-L185C86), we should design this a little bit differently.
E.g. user should be able to use `tune` API with just Job: `resources_per_trial={"gpu": 1}`, and with PyTorchJob.

Maybe we should create a new class which can be set in resources per Trial parameter:
```
DistributedResources(num_workers=1, num_procs_per_worker=5, resources_per_worker={"gpu": 5})
```
@droctothorpe @helenxie-bit @kubeflow/wg-training-leads Do you have any other ideas ?

Reply:
I agree with you. It's better to define a new class. I have updated it in the proposal.